MetaMask/metamask-mobile
# Refactor sanitization middleware

**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

**Description**

Add sanitiziation RPC middleware to process requests just before they get sent into our internal network provider. This is meant to be a refactor with no functional changes; this middleware is identical to [the middleware already used today within `web3-provider-engine`](https://github.com/MetaMask/web3-provider-engine/blob/9ac79a46ddbe37612027b08e22f395fd7dfeef04/subproviders/sanitizer.js).

This is being added so that we can remove `web3-provider-engine` with fewer functional changes.

**Test Instructions**

This should have no functional impact. This middleware is expected to process any RPC request not intercepted earlier in the pipeline however (essentially any request that gets sent to the RPC endpoint). It should ignore most requests (as the unit tests demonstrate), but it will process any that have an object as the first parameter. Known examples are `eth_call`, `eth_estimateGas`, and `eth_getLogs`.

To manually test this, you could try invoking these methods with parameters that the middleware should be expected to filter out. Or invoke them with parameters that are allowed, to confirm they remain unaffected. You could also try testing requests that we anticipate are unaffected. In all cases, we should see the same behavior before and after.

The test dapp has a new "Request" page that might be useful in testing this. See the README Usage section for details: https://github.com/MetaMask/test-dapp

**Issue**

Closes #5846

**Checklist**
